samza-core/src/main/scala/org/apache/samza/system/StreamMetadataCache.scala

```diff
@@ -39,10 +41,26 @@ class StreamMetadataCache (
     /** Clock used for determining expiry (for mocking in tests) */
     clock: Clock = SystemClock.instance) extends Logging {
 
-  private case class CacheEntry(metadata: SystemStreamMetadata, lastRefreshMs: Long)
-  private var cache = Map[SystemStream, CacheEntry]()
+  /**
+    * @param lastRefreshMs last time that the entry was updated
+    * @tparam T type of data in entry
+    */
+  private case class CacheEntry[T](data: T, lastRefreshMs: Long)
+
+  private var fullMetadataCache = Map[SystemStream, CacheEntry[SystemStreamMetadata]]()
   private val lock = new Object
 
+  /**
+    * Caches newest offset data.
+    * If there is a CacheEntry with Some data, and the TTL has not passed, then use the data and do not fetch.
+    * If there is a CacheEntry with None data, and the TTL has not passed, then use the empty entry and do not fetch.
+    * If there is no CacheEntry, then fetch the metadata and fill in the cache.
+    *
+    * Using a mutable.Map here instead of immutable.Map because this will get updated relatively often, so we do not
```

Comment:
Also, would it be reasonable to replace the lock + Map mutation with a ConcurrentHashMap instead?

Reply:
I initially was using mutable.Map for both, but then I realized that this is not threadsafe given the way that the fullMetadataCache is currently being read/written. I think immutable.Map gives thread safety since it creates a new Map on every update, so you can't catch it in an intermediate update state. I could try to add some thread safety protection around the fullMetadataCache if you would prefer that.
I don't think ConcurrentHashMap helps because if two threads both read a stale entry, then they will both try to fetch the metadata, and that defeats the caching benefit. I don't think ConcurrentHashMap can support atomically doing that refill. There is probably some actual threadsafe caching library we could use which provides TTL and a computeIfAbsent. I could look into that if you think it is worth it.